File: Neetcode/BinarySearch/BinarySearch.cpp
// Link: https://leetcode.com/problems/binary-search/
// Difficulty: Easy
// Date solved: May 6, 2023
/*
Solution
This is the binary search algorithm. The input is sorted so you can eliminate half of the array on each iteration.
You have a left and a right pointer, calculate the mid, and then compare it to the target. If the mid is the target,
you have found your value, if the mid is less than the target, the target will be to the right of mid, the same goes 
for if it is smaller. You can keep iterating until left <= right and if this fails then the number is not found and you
can return -1
*/

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if(nums[mid] == target) {
                return mid;
            }
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> test1 = {-1,0,3,5,9,12};
    cout << sol.search(test1, 9) << " should be 4" << endl;
    cout << sol.search(test1, 2) << " should be -1" << endl;
    return 0;
}

File: Neetcode/BinarySearch/Python/BinarySearch.py
# Link: https://leetcode.com/problems/binary-search/
# Difficulty: Easy
# Date Solved: Oct 5, 2023
# Solution
# Just the classical binary search. Init our left and right pointers
# and calcualte the midpointe, check if the mid is the value and if
# not move the left and right pointers one past or one behind the 
# midpoint

class Solution:
    def search(self, nums, target):
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1
